src/mod/endpoints/mod_sofia/sofia_reg.c

```diff
@@ -312,15 +312,19 @@ void sofia_reg_check_gateway(sofia_profile_t *profile, time_t now)
 	int delta = 0;
 
 	switch_mutex_lock(profile->gw_mutex);
+	switch_mutex_lock(profile->gw_deleting_mutex);
 	for (gateway_ptr = profile->gateways; gateway_ptr; gateway_ptr = gateway_ptr->next) {
 		if (gateway_ptr->deleted) {
 			if ((check = switch_core_hash_find(mod_sofia_globals.gateway_hash, gateway_ptr->name)) && check == gateway_ptr) {
 				char *pkey = switch_mprintf("%s::%s", profile->name, gateway_ptr->name);
 				switch_assert(pkey);
+				gateway_ptr->destroy = 1;
 				switch_log_printf(SWITCH_CHANNEL_LOG, SWITCH_LOG_DEBUG, "Removing gateway %s from hash.\n", pkey);
 				switch_core_hash_delete(mod_sofia_globals.gateway_hash, pkey);
 				switch_core_hash_delete(mod_sofia_globals.gateway_hash, gateway_ptr->name);
 				free(pkey);
+				switch_core_destroy_memory_pool(&(gateway_ptr->pool));
+				gateway_ptr->destroy = 0;
```

Comment:
We don't need to reset destroy flag here

Reply:
makes sense